### PR TITLE
Spanner plugin e2e test scenarios

### PR DESCRIPTION
**Pull request by sawantpritam**
@itsankit-google @rmstar Please review.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -885,8 +885,8 @@
           </includes>
           <!--Start configuration to run TestRunners in parallel-->
           <parallel>classes</parallel> <!--Running TestRunner classes in parallel-->
-          <threadCount>2</threadCount> <!--Number of classes to run in parallel-->
-          <forkCount>2</forkCount> <!--Number of JVM processes -->
+          <threadCount>1</threadCount> <!--Number of classes to run in parallel-->
```
Comment (rmstar) [inline pom.xml]: are we running into OOM issues with 2 parallel runs?

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 2]
```diff
@@ -214,4 +229,107 @@ public static void deleteTargetPubSubTopic() {
...
     }
   }
+
+  /**
+   * Create Spanner Instance with 1 database containing sample tables
+   * with queries provided in file spannerTestDataCreateTableQueriesFile and spannerTestDataInsertDataQueriesFile.
+   */
+  @Before(order = 1, value = "@SPANNER_TEST")
+  public static void createTempSpannerInstance() throws InterruptedException, ExecutionException {
+    if (firstSpannerTestFlag) {
```
Comment (rmstar) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java:275]: Is it guaranteed that this method won't be called concurrently? If not, you'll need some synchronization

File: src/e2e-test/java/io/cdap/plugin/spanner/stepsdesign/SpannerBase.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
```
Comment (rmstar) [inline src/e2e-test/java/io/cdap/plugin/spanner/stepsdesign/SpannerBase.java]: change to 2022 in all files

**Review by rmstar — COMMENTED** [review]
Are we testing all supported data types in both the source and sink plugin?

File: pom.xml (hunk 1 above)
Comment (sawantpritam) [reply pom.xml]: Yes we were facing issue for Spanner tests....Memory limit is increased. No issues with  2 threads now.

File: src/e2e-test/java/io/cdap/plugin/spanner/stepsdesign/SpannerBase.java (hunk 3 above)
Comment (sawantpritam) [reply src/e2e-test/java/io/cdap/plugin/spanner/stepsdesign/SpannerBase.java]: Changed to 2022

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java (hunk 2 above)
Comment (sawantpritam) [reply src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java:275]: Yes won't be called concurrently.

File: src/e2e-test/resources/testdata/SpannerCreateTableQueries.txt [hunk 4]
```diff
@@ -0,0 +1,2 @@
+CREATE TABLE EmployeeDetails ( EmployeeDepartment STRING(1024), EmployeeName STRING(1024), Salary INT64, Workhours INT64, DateOfBirth DATE, TimeOfBirth STRING(1024), AgeInYears FLOAT64 ) PRIMARY KEY(EmployeeDepartment, EmployeeName)
+CREATE TABLE Singers ( RowId INT64 NOT NULL, FirstName STRING(1024), LastName STRING(1024), BirthDate DATE, LastUpdated TIMESTAMP, ActiveRecord BOOL, CostOfItem NUMERIC, CostOfItem1 FLOAT64, RecordValue JSON ) PRIMARY KEY(RowId)
```
Comment (rmstar) [inline src/e2e-test/resources/testdata/SpannerCreateTableQueries.txt]: Are we using this table anywhere? Also, following up on a previous comment - can you please make sure you test all supported data types?

Comment (sawantpritam) [reply src/e2e-test/resources/testdata/SpannerCreateTableQueries.txt]: Updated EmployeeDetails with all supported data types.
Removed unused table Singers.

File: src/e2e-test/java/io/cdap/plugin/utils/SpannerClient.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * Represents GCP Spanner Client.
+ */
+public class SpannerClient {
+
+  public static int countBqQuery(String instanceId, String databaseId, String table) {
```
Comment (rmstar) [inline src/e2e-test/java/io/cdap/plugin/utils/SpannerClient.java]: Which is this named countBqQuery? This is not a BigQuery table

File: src/e2e-test/java/io/cdap/plugin/spanner/stepsdesign/SpannerSource.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import io.cdap.plugin.common.stepsdesign.TestSetupHooks;
+import io.cdap.plugin.spanner.actions.SpannerActions;
+import io.cucumber.java.en.Then;
+import io.cucumber.java.en.When;
+
+/**
+ * Spanner source plugin related test step definitions.
+ */
+public class SpannerSource implements CdfHelper {
+
```
Comment (rmstar) [inline src/e2e-test/java/io/cdap/plugin/spanner/stepsdesign/SpannerSource.java:29]: Indent looks incorrect. You can use [this](https://cdap.atlassian.net/wiki/spaces/CE/pages/1594098619/Coding+Standard) Intellij settings file. Please reformat all files in PR.

**Review by rmstar — COMMENTED** [review]
Are we testing all supported data types in spanner sink?

File: src/e2e-test/java/io/cdap/plugin/utils/SpannerClient.java (hunk 5 above)
Comment (sawantpritam) [reply src/e2e-test/java/io/cdap/plugin/utils/SpannerClient.java]: renamed method.

File: src/e2e-test/java/io/cdap/plugin/spanner/stepsdesign/SpannerSource.java (hunk 6 above)
Comment (sawantpritam) [reply src/e2e-test/java/io/cdap/plugin/spanner/stepsdesign/SpannerSource.java:29]: Indent correction done.

**Comment by sawantpritam** [comment]
> Are we testing all supported data types in spanner sink?

We will add in separate PR. https://cdap.atlassian.net/browse/CDAP-18908

**Review by rmstar — APPROVED** [review]
please squash commits

**Comment by sawantpritam** [comment]
Squashed commits.

**Comment by itsankit-google** [comment]
@sawantpritam Could you PTAL why some tests are failing?

**Comment by sawantpritam** [comment]
Rebased.

**Pull request merged**